pkg/ingester/ingester.go

```diff
@@ -495,6 +536,58 @@ func (i *Ingester) append(ctx context.Context, userID string, labels labelPairs,
 	return err
 }
 
+func (i *Ingester) appendMetadata(ctx context.Context, userID string, metadata *client.MetricMetadata) error {
+	i.userStatesMtx.RLock()
+	if i.stopped {
+		i.userStatesMtx.RUnlock()
+		return fmt.Errorf("ingester stopping")
+	}
+	i.userStatesMtx.RUnlock()
+
+	i.usersMetadataMtx.Lock()
+	defer i.usersMetadataMtx.Unlock()
```

Comment:
In my idea (the designed I shared with you and Peter offline) I suggested to use this lock only for the `getOrCreateUserMetadata()` and then having a lock inside `metadataMap struct` basically making `metadataMap` and `memoryMetadataSet` functions goroutines safe. This way we would have a fine-grained lock instead of a global lock.

The global lock here guarantees perfect consistency in the code below (ie. between `i.getOrCreateUserMetadata()` and `userMetadata.put()`) but I think we don't need a perfect limit while a global lock is quite scary.

@pstibrany How do you feel with a global lock here?